lib/styleToFilters.js

```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
           layers[layerName].minzoom = style.layers[i].minzoom || 0;
           layers[layerName].maxzoom = style.layers[i].maxzoom || 22;
         }
+
+        // Collect the used property
+        layers[layerName].properties = layers[layerName].properties || [];
+        ['paint', 'layout'].forEach(item => {
+          let itemObject = style.layers[i][item];
+          itemObject && getPropertyLoop(itemObject, layers[layerName].properties);
+        });
       }
     }
   }
 
+  // remove duplicate propertys and fix choose all the propertys in layers[i].properties
+  Object.keys(layers).forEach(layerId => {
+    let properties = layers[layerId].properties;
+    if (properties.indexOf(true) !== -1) {
+      layers[layerId].properties = true;
+    } else {
+      let unique = {};
+      properties.forEach(function(i) {
+        if (!unique[i]) {
+          unique[i] = true;
+        }
+      });
+      layers[layerId].properties = Object.keys(unique);
+    }
+  })
+
   return layers;
 }
 
+
+function getPropertyLoop(propertyObj, properties) {
+  Object.keys(propertyObj).forEach(key => {
+    let value = propertyObj[key];
+    // TODO we still have outher situation:
+    // - special properties: `mapbox_clip_start`, `mapbox_clip_end`
+    if (typeof value === 'string') {
+      // if the value is string try to get property name from `xx{PropertyName}xx` like.
+      // the /{[^}]+}/ig return all the value like {xxx}
+      // eg 'a{hello}badfa' => ['{hello}']
+      // eg 'a{hello}ba{world}dfa' => ['{hello}','{world}']
+      let preProperties = value.match(/{[^}]+}/ig);
+      preProperties && preProperties.forEach(item => {
+        properties.push(item.slice(1, -1));
+      });
+    } else if (typeof value === 'object' && typeof value.property === 'string') {
+      // - legacy functions with `property`
+      properties.push(value.property);
+    } else {
+      // test isExpression from sytleSpec
+      if (styleSpec.expression.isExpression(value)) {
+        // TODO: now we implement this by ourself in vtshavem, we need to talk with ‘style spec’ member to see if there have a official method to get used property, to make this can be synchronized with the expression update.
+        getPropertyFromExpression(value, properties);
+      } else {
+        // otherwise continual loop;
+        getPropertyLoop(value, properties);
+      }
+    }
+  })
+}
+
+
+function getPropertyFromExpression(exp, properties) {
+  // now we care about the expression like:
+  // ["get", string] not ["get", string, Object],
+  // ["has", string] not ["has", string, Object],
+  // ["properties"],
+  // ["feature-state", string]
```

Comment:
Are these code comments necessary now that we have unit tests? If not I'd recommend removing them.

Reply:
Yeah, unit tests can tell if all the condition are right. 
One of my concern is that I should let the code reader or the "6 month later me" can understand what we are doing here, and how many codition we need to deal with, below this comment are some trick code, with out the comment it's hard to understand, so I write this.
I think we should keep this, how do you think?

Reply:
> I think we should keep this, how do you think?

Okay, 👍 to keeping. Can I ask that you make the comments more detailed then. It is hard for me to know what `["has", string] not ["has", string, Object],` means. I think it would be good to explain, in comments, with more detail.

Reply:
yeah, I'll add the more detail of the comment.

Reply:
👌